content/blog/permutate-parsers/index.md

```diff
@@ -0,0 +1,814 @@
+```
+
+Here is a new batch for us to peruse:
+
+```plain
+pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980
+hcl:#623a2f
+
+eyr:1972 cid:100
+hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926
+```
+
+- the first passport is _valid_
+- the second passport is _invalid_ (look at the `eyr` field)
+
+These new requirements are a bit annoying. Our simple approach of checking
+if all required fields are present won't work any longer. We can instead
+implement a `isFieldValid` function to check if all fields are valid.
+
+```haskell
+isFieldValid :: (PassportField, String) -> Bool
+isFieldValid (field, value) =
+  case field of
+    BirthYear ->
+      let v = toInt value
+       in length value == 4 && v >= 1920 && v <= 2002
+    IssueYear ->
+      let v = toInt value
+       in length value == 4 && v >= 2010 && v <= 2020
+    ExpirationYear ->
+      let v = toInt value
+       in length value == 4 && v >= 2020 && v <= 2030
+    Height ->
+      case span Char.isDigit value of
+        (num, "cm") ->
+          let n = toInt num
+           in n >= 150 && n <= 193
+        (num, "in") ->
+          let n = toInt num
+           in n >= 59 && n <= 76
+        _ ->
+          False
+    HairColor ->
+      case (length value, value) of
+        (7, '#' : rest) ->
+          all (`elem` allowedHexChars) rest
+        _ ->
+          False
+    EyeColor ->
+      value `elem` validEyeColors
+    PassportId ->
+      length value == 9 && all Char.isDigit value
+    CountryId ->
+      all Char.isDigit value
+
+toInt :: String -> Int
+toInt = read
+
+validEyeColors :: [String]
+validEyeColors =
+  ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
+
+allowedHexChars :: [Char]
+allowedHexChars =
+  ['0' .. '9'] <> ['a' .. 'f']
+```
+
+Then we can change our `isEntryValid` to use this function:
+
+```haskell
+isEntryValid :: PassportEntry -> Bool
+isEntryValid entry =
+  requiredFieldsPresent && allFieldsValid
+  where
+    requiredFieldsPresent =
+      all (`HM.member` entry) requiredFields
+
+    allFieldsValid =
+      all isFieldValid (HM.toList entry)
+```
+
+Running this program on our second data sample yields `1`, and it will be
+good enough to solve the Advent of Code challenge and get you those sweet
+sweet stars.
+
+🎉 🎉 🎉
+
+## A moment of reflection
+
+If we look back at the [current
+state](https://gist.github.com/Arkham/923df6e872d855101bca4261f7d71e65) of
+our code, we can see that we are doing a lot of validations.
+
+We do a lot of work to verify if something is valid, then throw it all out
+of the window to return a meagre `Bool`. German folks from the [sixteenth
+century](https://en.wikipedia.org/wiki/Don%27t_throw_the_baby_out_with_the_bathwater)
+would have told us snarkily:
+
+> das Kind mit dem Bade ausschütten
+
+Yes, we're literally _throwing the baby out with the bathwater_.
+
+We realize this is even more true if we tried to take the data and fill
+that ideal `Passport` representation we were mentioning earlier:
```

Comment:
🎨  💭  For someone who isn't already sold on the idea of parsing over validating, it might not be evident that this representation is better than the hashmap, and why we lost useful information along the way.

You mention later that it's hard to get stuff out, but it may be nice to begin section with an example of how using the hash forces you to check things you already check to illustrate that and motivate what's coming.

Reply:
Agree, I've added a new question "Now find the unique set of eye colors in all valid passports" to showcase what happens when you just validate :)